[GTK] D-Bus proxy quietly fails if host bus address is not mounted in xdg-dbus-proxy's sandbox

mcatanzaro opened the pull request:
#### 67cda4acff9b343c48a2b24c326a2be08642ccee
<pre>
[GTK] D-Bus proxy quietly fails if host bus address is not mounted in xdg-dbus-proxy&apos;s sandbox
<a href="https://bugs.webkit.org/show_bug.cgi?id=246159">https://bugs.webkit.org/show_bug.cgi?id=246159</a>

Reviewed by Carlos Garcia Campos.

D-Bus 1.15.2 has changed the default session bus address to a filesystem
socket that lives under /tmp. However, our xdg-dbus-proxy cannot access
this location because we assume the session bus socket will always be
mounted under /run, since that&apos;s where all major distros put it. It&apos;s OK
to be flexible and mount absolutely any directory, whatever it may be,
since we&apos;re not actually trying to create a sandbox that the
xdg-dbus-proxy cannot break out of. It&apos;s a trusted process, and the
sandbox exists solely so that portals can verify the app ID of the
process that is using the proxy, which is done by inspecting
/.flatpak-info in its mount namespace&apos;s filesystem root. So let&apos;s mount
whatever directory is in use and move on. Credit to oreo639 for
investigating the problem and proposing a fix in WebKit#5011.

The a11y bus has the same theoretical problem, although it&apos;s not an
issue today because currently it will always be under /run in
practice. Still, we should fix it. There is one complication:
PlatformDisplay currently uses just one variable for both the host a11y
bus address and the proxy bus address, relying on XDGDBusProxy to change
it from the host address to the proxy address. This is fragile and it&apos;s
easier to fix it than to work around it by caching the value before it
changes, so at Carlos&apos;s suggestion, I have removed the ability to
overwrite the value in PlatformDisplay, and added a separate variable to
track the proxy address in WTF&apos;s Sandbox helpers.

I have snuck in a drive-by cleanup to avoid duplicating BASE_DIRECTORY
between two files, a problem that I introduced in 255218@main.
Additionally, I remove a stale declaration for XDGDBusProxy::makePath,
which I forgot to delete after removing the function in the same commit.

Finally, always add the extra sandbox paths to the sandbox. These were
originally extra paths for the web process only, but changed to be extra
paths for both web process and D-Bus proxy. It&apos;s no longer needed except
for the web process, but there&apos;s no particular reason to limit it
either. I&apos;m changing this here only because it&apos;s right next to the code
I&apos;m editing anyway, and it&apos;s odd to be adding extra sandbox paths
specifically for the D-Bus proxy process.

Canonical link: <a href="https://commits.webkit.org/255530@main">https://commits.webkit.org/255530@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/14ccf3e88d44de3761eb1e9dabe7f156cd62d263

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/92843 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/2057 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/23431 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/102570 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/162859 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/96845 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/2058 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/30397 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/85246 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/98725 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/98507 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/1404 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/79333 "Built successfully") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/28322 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/83317 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/83040 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/71440 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/84190 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/36801 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/16944 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/79244 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/34601 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/18134 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/27452 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3831 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/38471 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/40738 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/81872 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/40386 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/37309 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/18514 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebCore/platform/graphics/PlatformDisplay.h:
```diff
@@ -105,7 +105,8 @@ class PlatformDisplay {
 #if USE(ATSPI)
-    void setAccessibilityBusAddress(String&& address) { m_accessibilityBusAddress = WTFMove(address); }
+    void setAccessibilityBusProxyAddress(String&& address) { m_accessibilityBusProxyAddress = WTFMove(address); }
+    const String accessibilityBusProxyAddress() const;
```
I don't think this is right, the platform display only cares about the actual address to use, the same way we override the env var, whether it's the actual one or the socket one is a detail the display doesn't need to know.

mcatanzaro replied:
Hm, then I think I need to move all the code that computes the accessibility bus address to someplace else. Will consider this.

mcatanzaro replied:
I don't think PlatformDisplay really has anything to do with the accessibility bus at all. We've just put the code to compute the accessibility bus address here because it's a convenient global location, and it seems as nice a place for it to go as anywhere else does, so I'm not inclined to try moving it.

The problem is that XDGDBusProxy overrides the address here with an address pointing to the proxy socket instead, but we still need the original unmodified address later on when launching the xdg-dbus-proxy process. So the problem to solve is: do not overwrite the original address. I thought separating the original address from the proxy address was a nice way to do it, but any other way would work just fine. My next proposal will be a new function `PlatformDisplay::originalAccessibilityBusAddress` that works the same as `PlatformDisplay::accessibilityBusAddress` except it ignores the value set with `PlatformDisplay::setAccessibilityBusAddress`. That should work. I wanted to call it `PlatformDisplay::hostAccessibilityBusAddress`, but that would be confusing when running in the web process, because they'll return the same result in the web process, which never sees the real host bus address, only the proxy address.

mcatanzaro replied:
Updated, does this look OK to you now? And if not, do you have an alternative suggestion?

carlosgcampos replied:
The reason why it was added to PlatformDisplay is because in X11, the bus address is set as an atom in the x11 display (AT_SPI_BUS), see PlatformDisplayX11::platformAccessibilityBusAddress().

mcatanzaro commented on Source/WebKit/UIProcess/glib/WebProcessPoolGLib.cpp:
```diff
@@ -120,7 +120,7 @@ void WebProcessPool::platformInitializeWebProcess(const WebProcessProxy& process
 #if USE(ATSPI)
     static const char* accessibilityBusAddress = getenv("WEBKIT_A11Y_BUS_ADDRESS");
-    parameters.accessibilityBusAddress = accessibilityBusAddress ? String::fromUTF8(accessibilityBusAddress) : WebCore::PlatformDisplay::sharedDisplay().accessibilityBusAddress();
+    parameters.accessibilityBusAddress = accessibilityBusAddress ? String::fromUTF8(accessibilityBusAddress) : WebCore::PlatformDisplay::sharedDisplay().accessibilityBusProxyAddress();
```
I think this fails when the sandbox is disabled. Maybe that's why the API tests are failing?

nater1983 replied:
But thats the thing we shouldnt want the sandbox disable though. Not unless there is nothing using it at that moment.

mcatanzaro replied:
No, I mean the address will be empty when the sandbox is disabled, because I never set it.

We have to proxy access to the accessibility bus because unrestricted access would be a sandbox escape.

carlosgcampos commented on Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp:
```diff
@@ -686,16 +716,31 @@ GRefPtr<GSubprocess> bubblewrapSpawn(GSubprocessLauncher* launcher, const Proces
+                "--bind", *a11yBusDirectory, *a11yBusDirectory,
+            }));
+        }
+#endif
```
The problem of this approach is that in the case of wayland we end up doing two sync dbus calls to get the actual a11y bus address, once here in and one in XDGDBusProxy::accessibilityProxy(). So, I think using the PlatformDisplay to override the a11y bus address was not a good idea, because it's only done in the UI process which is not sandboxed, but only because we need a way to pass the sandboxed bus address to the web process via creation parameters. What I would do is:
- Remove the setAccessibilityBusAddress from PlatformDisplay: the display always returns the original bus address and it's cached to ensure there's only one sync dbus call.
- Add setSandboxedAccessibilityBusAddress() and sandboxedAccessibilityBusAddress() to Sandbox.h in WTF
- Use WTF Sandboxed API to set and get the global sandboxed a11y bus in the UI process.

What do you think?

mcatanzaro replied:
OK, that ought to work.

mcatanzaro replied:
I've implemented your suggestion. Does it look OK?

carlosgcampos commented on Source/WebKit/UIProcess/glib/WebProcessPoolGLib.cpp:
```diff
@@ -119,8 +120,16 @@ void WebProcessPool::platformInitializeWebProcess(const WebProcessProxy& process
+
+        String& address = sandboxedAccessibilityBusAddress();
+        if (!address.isNull())
+            return address;
```
I know this is an existing issue, but I've realized now. Even if it's not common case there might be two contexts, one with the sandbox enabled and the other disabled, so we should always check here if sandbox is enabled in this particular web process pool.

mcatanzaro replied:
Good catch!

carlosgcampos reviewed (approved):
Please add m_sandboxEnabled check before landing.

mcatanzaro committed:
[GTK] D-Bus proxy quietly fails if host bus address is not mounted in xdg-dbus-proxy's sandbox
https://bugs.webkit.org/show_bug.cgi?id=246159

Reviewed by Carlos Garcia Campos.

D-Bus 1.15.2 has changed the default session bus address to a filesystem
socket that lives under /tmp. However, our xdg-dbus-proxy cannot access
this location because we assume the session bus socket will always be
mounted under /run, since that's where all major distros put it. It's OK
to be flexible and mount absolutely any directory, whatever it may be,
since we're not actually trying to create a sandbox that the
xdg-dbus-proxy cannot break out of. It's a trusted process, and the
sandbox exists solely so that portals can verify the app ID of the
process that is using the proxy, which is done by inspecting
/.flatpak-info in its mount namespace's filesystem root. So let's mount
whatever directory is in use and move on. Credit to oreo639 for
investigating the problem and proposing a fix in WebKit#5011.

The a11y bus has the same theoretical problem, although it's not an
issue today because currently it will always be under /run in
practice. Still, we should fix it. There is one complication:
PlatformDisplay currently uses just one variable for both the host a11y
bus address and the proxy bus address, relying on XDGDBusProxy to change
it from the host address to the proxy address. This is fragile and it's
easier to fix it than to work around it by caching the value before it
changes, so at Carlos's suggestion, I have removed the ability to
overwrite the value in PlatformDisplay, and added a separate variable to
track the proxy address in WTF's Sandbox helpers.

I have snuck in a drive-by cleanup to avoid duplicating BASE_DIRECTORY
between two files, a problem that I introduced in 255218@main.
Additionally, I remove a stale declaration for XDGDBusProxy::makePath,
which I forgot to delete after removing the function in the same commit.

Finally, always add the extra sandbox paths to the sandbox. These were
originally extra paths for the web process only, but changed to be extra
paths for both web process and D-Bus proxy. It's no longer needed except
for the web process, but there's no particular reason to limit it
either. I'm changing this here only because it's right next to the code
I'm editing anyway, and it's odd to be adding extra sandbox paths
specifically for the D-Bus proxy process.

Canonical link: https://commits.webkit.org/255530@main
